File: App/frontend/src/toolCall/validation/validators.ts
/**
 * Validators
 *
 * Individual validator functions for tool call validation.
 * Validators access stores directly via getState() for async operations.
 */

import { useUnifiedObjectStore } from '../../store/unifiedObjectStore';
import { useSettingsStore } from '../../store/settingsStore';
import type { ObjectType } from '../../types/unifiedObject';
import type { Validator } from './types';
import { validResult, invalidResult } from './types';
import { STORY_OBJECT_TYPE_MAP, ALL_OBJECT_TYPE_MAP, type StoryObjectSubtype } from '../types';
import { applySingleReplacement } from '../../utils/patchUtils';
import { getObjectData } from '../types';
import { docToMarkdown } from '../../editor/manuscript/convert';
import { normalizeDoc } from '../../editor/manuscript/doc';
import { toolSchemaResolver } from '../schemas/toolSchemaResolver';

// ============================================================================
// HELPER FUNCTIONS
// ============================================================================

/**
 * Resolve ObjectType from function arguments and function name
 */
export function resolveObjectType(
  args: Record<string, unknown>,
  toolName: string
): ObjectType | null {
  const type = args.type as string | undefined;

  // Read functions - resolve from args.type
  if (toolName === 'read_story_object' && type) {
    // read_story_object can read basic_info and guidelines too
    if (type === 'basic_info') return 'basic_info';
    if (type === 'guidelines') return 'guidelines';
    return STORY_OBJECT_TYPE_MAP[type as StoryObjectSubtype] ?? null;
  }
  if (toolName === 'read_outline' && type) {
    // type is 'outline' | 'act' | 'chapter'
    if (type === 'outline' || type === 'act' || type === 'chapter') {
      return type as ObjectType;
    }
    return null;
  }
  if (toolName === 'read_manuscript') {
    return 'manuscript';
  }

  // Story object functions
  if (toolName.includes('story_object') && type) {
    return STORY_OBJECT_TYPE_MAP[type as StoryObjectSubtype] ?? null;
  }

  // Outline functions
  if (toolName.includes('outline_act')) {
    return 'act';
  }
  if (toolName.includes('outline_chapter')) {
    return 'chapter';
  }
  if (toolName === 'patch_outline' || toolName === 'replace_outline' || toolName === 'delete_outline') {
    return 'outline';
  }

  // Basic info
  if (toolName.includes('basic_info')) {
    return 'basic_info';
  }

  // Guidelines
  if (toolName.includes('guidelines')) {
    return 'guidelines';
  }

  // Manuscript
  if (toolName.includes('manuscript')) {
    return 'manuscript';
  }

  // Translation with type argument
  if (toolName.includes('_translation') && type) {
    return ALL_OBJECT_TYPE_MAP[type] ?? null;
  }

  return null;
}

// ============================================================================
// SYNC VALIDATORS
// ============================================================================

/**
 * Validate tool call against its schema (known tool + required args + enums).
 *
 * This runs for every tool call (including dynamic call_{agent_name} tools).
 */
export const validateToolSchema: Validator = (args, toolName, _context) => {
  const schema = toolSchemaResolver.get(toolName);
  if (!schema) {
    return invalidResult(`Unknown tool: ${toolName}`);
  }

  const errors: string[] = [];
  const required = Array.isArray((schema.parameters as any)?.required) ? (schema.parameters as any).required : [];
  const properties = (schema.parameters as any)?.properties ?? {};

  for (const param of required) {
    const value = (args as any)[param];
    if (value === undefined || value === null) {
      errors.push(`Missing required parameter: ${param}`);
      continue;
    }
    if (typeof value === 'string' && !value.trim()) {
      errors.push(`Required parameter cannot be empty: ${param}`);
    }
  }

  for (const [key, value] of Object.entries(args)) {
    const propSchema = properties?.[key] as Record<string, unknown> | undefined;
    const enumValues = propSchema?.enum;
    if (Array.isArray(enumValues) && !enumValues.includes(value)) {
      errors.push(`Invalid value for ${key}: ${String(value)}. Expected one of: ${enumValues.join(', ')}`);
    }
  }

  if (errors.length > 0) {
    return invalidResult(errors.join(', '));
  }

  return validResult();
};

/**
 * Validate that the tool was actually provided to the model for this session.
 * (Prevents the model from calling tools outside the current allowlist.)
 */
export const validateToolAllowedInSession: Validator = (_args, toolName, context) => {
  const allowed = context.allowedToolNames;
  if (!allowed || allowed.length === 0) return validResult();
  if (allowed.includes(toolName)) return validResult();
  return invalidResult(`Tool not available in this session: ${toolName}`);
};

/**
 * Validate dynamic Sub Agent call tools: call_{agent_name}({ input: string }).
 */
export const validateSubAgentCallInput: Validator = (args, toolName, _context) => {
  if (!toolName.startsWith('call_')) return validResult();

  const input = (args as any).input;
  if (typeof input !== 'string' || !input.trim()) {
    return invalidResult('call_* tools require a non-empty string field "input".');
  }

  return validResult();
};

/**
 * Validate that the type argument maps to a valid ObjectType
 */
export const validateTypeMapping: Validator = (args, toolName, _context) => {
  const type = args.type as string | undefined;

  // Only validate if type argument is present
  if (!type) {
    return validResult();
  }

  // Check story object type mapping
  if (toolName.includes('story_object')) {
    if (!STORY_OBJECT_TYPE_MAP[type as StoryObjectSubtype]) {
      return invalidResult(`Unknown story object type: ${type}`);
    }
  }

  // Check translation object type mapping (includes all object types)
  if (toolName.includes('_translation')) {
    if (!ALL_OBJECT_TYPE_MAP[type]) {
      return invalidResult(`Unknown object type for translation: ${type}`);
    }
  }

  return validResult();
};

/**
 * Validate that required ID field is present
 */
export const validateRequiredId: Validator = (args, toolName, _context) => {
  const id = args.id as string | undefined;

  // Create operations don't require ID
  if (toolName.startsWith('create_')) {
    return validResult();
  }

  // All other operations require ID
  if (!id) {
    return invalidResult(`Missing required id for ${toolName}`);
  }

  return validResult();
};

/**
 * Validate required fields for patch operations
 */
export const validatePatchRequiredFields: Validator = (args, toolName, _context) => {
  if (!toolName.startsWith('patch_')) {
    return validResult();
  }

  const oldText = args.old as string | undefined;
  const newText = args.new as string | undefined;

  if (!oldText || newText === undefined) {
    return invalidResult(`Missing required fields (old, new) for ${toolName}`);
  }

  // For non-manuscript patches, field is required
  if (!toolName.includes('manuscript')) {
    const field = args.field as string | undefined;
    if (!field) {
      return invalidResult(`Missing required field 'field' for ${toolName}`);
    }
  }

  return validResult();
};

/**
 * Validate that RAG Search is enabled for rag_search tool calls.
 */
export const validateRagSearchEnabled: Validator = (_args, toolName, _context) => {
  if (toolName !== 'rag_search') return validResult();

  const enabled = useSettingsStore.getState().settings.ragSearchEnabled;
  if (!enabled) {
    return invalidResult('Embeddings are disabled (Settings > RAG Search)');
  }

  return validResult();
};

/**
 * Validate that Sub Agent return result contains a non-empty string payload.
 *
 * Note: This only validates arguments. Cross-tool rules (e.g., "must be last and alone")
 * are enforced at the Sub Agent orchestration layer.
 */
export const validateReturnSubAgentResultArgs: Validator = (args, toolName, _context) => {
  if (toolName !== 'return_sub_agent_result') return validResult();

  const result = args.result;
  if (typeof result !== 'string' || !result.trim()) {
    return invalidResult(
      'return_sub_agent_result requires a non-empty string field "result". ' +
      'Call it exactly once, and only as the final, standalone tool call in the Sub Agent invocation.'
    );
  }

  return validResult();
};

// ============================================================================
// ASYNC VALIDATORS
// ============================================================================

/**
 * Validate that the object exists in the store (fetching if needed)
 */
export const validateObjectExists: Validator = async (args, toolName, context) => {
  const id = args.id as string | undefined;

  // Create operations don't need existing object
  if (toolName.startsWith('create_') || !id) {
    return validResult();
  }

  const store = useUnifiedObjectStore.getState();

  // Check if already in cache
  let object = store.getObject(id);

  if (!object) {
    // Try to fetch from API
    const objectType = resolveObjectType(args, toolName);
    if (!objectType) {
      return invalidResult(`Cannot determine object type for ${toolName}`);
    }

    try {
      await store.fetchObject(objectType, id);
      object = store.getObject(id);
    } catch (error) {
      return invalidResult(`Failed to fetch object ${id}: ${error}`);
    }
  }

  if (!object) {
    return invalidResult(`Object with id ${id} not found`);
  }

  const objectProjectId = object.metadata?.project_id;
  if (!objectProjectId) {
    return invalidResult(`Object ${id} is missing project_id metadata`);
  }

  if (objectProjectId !== context.projectId) {
    return invalidResult(`Object ${id} belongs to a different project`);
  }

  return validResult();
};

/**
 * Validate that patch replacement can be applied to current content.
 * Uses applySingleReplacement as a dry run to check if the patch will succeed.
 */
export const validatePatchApplicable: Validator = async (args, toolName, context) => {
  if (!toolName.startsWith('patch_')) {
    return validResult();
  }

  const id = args.id as string | undefined;
  const oldText = args.old as string | undefined;
  const newText = args.new as string | undefined;
  const field = args.field as string | undefined;

  // For basic_info, id is not required
  if (toolName !== 'patch_basic_info' && !id) {
    return validResult(); // Already validated by other validators
  }

  if (!oldText || newText === undefined) {
    return validResult(); // Already validated by validatePatchRequiredFields
  }

  const store = useUnifiedObjectStore.getState();

  // For basic_info, we need to get the object differently
  if (toolName === 'patch_basic_info') {
    const existing = await store.listObjects('basic_info', context.projectId);
    if (existing.length === 0) {
      return invalidResult('No basic info found to patch');
    }
    const object = existing[0];
    const currentData = getObjectData(object, context.language);
    const fieldValue = (currentData[field as string] as string) ?? '';
    const result = applySingleReplacement(fieldValue, oldText, newText);
    if (!result.success) {
      return invalidResult(`Patch validation failed: [${field}] ${result.error}`);
    }
    return validResult();
  }

  const object = store.getObject(id as string);
  if (!object) {
    return validResult(); // Already validated by validateObjectExists
  }

  const currentData = getObjectData(object, context.language);

  // For manuscript patches (no field property), validate against content
  if (toolName.includes('manuscript')) {
    const currentDoc = normalizeDoc((currentData as any).doc);
    const markdown = docToMarkdown(currentDoc);
    const result = applySingleReplacement(markdown, oldText, newText);
    if (!result.success) {
      return invalidResult(`Patch validation failed: ${result.error}`);
    }
  } else {
    // For other patches with field property
    if (!field) {
      return invalidResult('Missing field for patch validation');
    }
    const fieldValue = (currentData[field] as string) ?? '';
    const result = applySingleReplacement(fieldValue, oldText, newText);
    if (!result.success) {
      return invalidResult(`Patch validation failed: [${field}] ${result.error}`);
    }
  }

  return validResult();
};

// ============================================================================
// OUTLINE PARENT VALIDATORS
// ============================================================================

export const validateOutlineParentExists: Validator = async (args, toolName, context) => {
  if (toolName !== 'create_outline_act') {
    return validResult();
  }

  const outlineId = args.outlineId as string | undefined;
  if (!outlineId || !outlineId.trim()) {
    return invalidResult('Missing outlineId for create_outline_act');
  }

  const store = useUnifiedObjectStore.getState();

  let outline = store.getObject(outlineId);
  if (!outline) {
    try {
      await store.fetchObject('outline', outlineId);
      outline = store.getObject(outlineId);
    } catch (error) {
      return invalidResult(`Failed to fetch outline ${outlineId}: ${error instanceof Error ? error.message : String(error)}`);
    }
  }

  if (!outline) {
    return invalidResult(`Parent outline not found: ${outlineId}`);
  }

  const projectId = outline.metadata?.project_id;
  if (!projectId) {
    return invalidResult(`Parent outline ${outlineId} is missing project_id metadata`);
  }

  if (projectId !== context.projectId) {
    return invalidResult(`Parent outline ${outlineId} belongs to a different project`);
  }

  return validResult();
};

export const validateActParentExists: Validator = async (args, toolName, context) => {
  if (toolName !== 'create_outline_chapter') {
    return validResult();
  }

  const actId = args.actId as string | undefined;
  if (!actId || !actId.trim()) {
    return invalidResult('Missing actId for create_outline_chapter');
  }

  const store = useUnifiedObjectStore.getState();

  let act = store.getObject(actId);
  if (!act) {
    try {
      await store.fetchObject('act', actId);
      act = store.getObject(actId);
    } catch (error) {
      return invalidResult(`Failed to fetch act ${actId}: ${error instanceof Error ? error.message : String(error)}`);
    }
  }

  if (!act) {
    return invalidResult(`Parent act not found: ${actId}`);
  }

  const projectId = act.metadata?.project_id;
  if (!projectId) {
    return invalidResult(`Parent act ${actId} is missing project_id metadata`);
  }

  if (projectId !== context.projectId) {
    return invalidResult(`Parent act ${actId} belongs to a different project`);
  }

  return validResult();
};

File: App/backend/services/validators.py
from __future__ import annotations

from dataclasses import dataclass
import inspect
from typing import Any, Awaitable, Callable
from uuid import UUID

from sqlalchemy.orm import Session

from ..models.db_models import (
    Act,
    BasicInfo,
    Chapter,
    Guidelines,
    Character,
    Location,
    LorebookEntry,
    Manuscript,
    Organization,
    Outline,
    RunModel,
)
from ..models.translation_models import ObjectVersion
from .patch_utils import apply_single_replacement
from .sidecar_client import SidecarClient, SidecarConversionError, SidecarUnavailableError
from .tool_schemas import ToolSchemaDef, get_tool_schema


STORY_TYPE_MAP: dict[str, str] = {
    "character": "character",
    "location": "location",
    "organization": "organization",
    "lorebook": "lorebook",
}

ALL_OBJECT_TYPE_MAP: dict[str, str] = {
    **STORY_TYPE_MAP,
    "basic_info": "basic_info",
    "guidelines": "guidelines",
    "outline": "outline",
    "act": "act",
    "chapter": "chapter",
    "manuscript": "manuscript",
}


@dataclass
class ValidationResult:
    valid: bool
    reason: str | None = None
    validator: str | None = None


@dataclass
class ValidationContext:
    db: Session
    run: RunModel
    allowed_tool_names: set[str] | None
    schema_by_name: dict[str, ToolSchemaDef]
    rag_search_enabled: bool
    sidecar: SidecarClient | None = None


Validator = Callable[[dict[str, Any], str, ValidationContext], ValidationResult | Awaitable[ValidationResult]]


def valid_result() -> ValidationResult:
    return ValidationResult(valid=True)


def invalid_result(validator: str, reason: str) -> ValidationResult:
    return ValidationResult(valid=False, reason=reason, validator=validator)


def _schema_for(tool_name: str, ctx: ValidationContext) -> ToolSchemaDef | None:
    if tool_name in ctx.schema_by_name:
        return ctx.schema_by_name[tool_name]
    return get_tool_schema(tool_name)


def _resolve_object_type(args: dict[str, Any], tool_name: str) -> str | None:
    raw_type = args.get("type")
    type_value = raw_type if isinstance(raw_type, str) else None

    if tool_name == "read_story_object" and type_value:
        if type_value in {"basic_info", "guidelines"}:
            return type_value
        return STORY_TYPE_MAP.get(type_value)

    if tool_name == "read_outline" and type_value:
        if type_value in {"outline", "act", "chapter"}:
            return type_value
        return None

    if tool_name == "read_manuscript":
        return "manuscript"

    if "story_object" in tool_name and type_value:
        return STORY_TYPE_MAP.get(type_value)

    if "outline_act" in tool_name:
        return "act"
    if "outline_chapter" in tool_name:
        return "chapter"
    if tool_name in {"patch_outline", "replace_outline", "delete_outline"}:
        return "outline"

    if "basic_info" in tool_name:
        return "basic_info"
    if "guidelines" in tool_name:
        return "guidelines"
    if "manuscript" in tool_name:
        return "manuscript"

    if type_value in ALL_OBJECT_TYPE_MAP:
        return ALL_OBJECT_TYPE_MAP[type_value]
    return None


def _parse_required_field(args: dict[str, Any], name: str) -> str | None:
    value = args.get(name)
    if not isinstance(value, str):
        return None
    return value.strip() or None


def _load_object_for_project(db: Session, *, object_type: str, object_id: UUID, project_id: UUID) -> Any | None:
    if object_type == "basic_info":
        return db.query(BasicInfo).filter(BasicInfo.id == object_id, BasicInfo.project_id == project_id).first()
    if object_type == "guidelines":
        return db.query(Guidelines).filter(Guidelines.id == object_id, Guidelines.project_id == project_id).first()
    if object_type == "character":
        return db.query(Character).filter(Character.id == object_id, Character.project_id == project_id).first()
    if object_type == "location":
        return db.query(Location).filter(Location.id == object_id, Location.project_id == project_id).first()
    if object_type == "organization":
        return db.query(Organization).filter(Organization.id == object_id, Organization.project_id == project_id).first()
    if object_type == "lorebook":
        return db.query(LorebookEntry).filter(LorebookEntry.id == object_id, LorebookEntry.project_id == project_id).first()
    if object_type == "outline":
        return db.query(Outline).filter(Outline.id == object_id, Outline.project_id == project_id).first()
    if object_type == "act":
        return (
            db.query(Act)
            .join(Outline, Outline.id == Act.outline_id)
            .filter(Act.id == object_id, Outline.project_id == project_id)
            .first()
        )
    if object_type == "chapter":
        return (
            db.query(Chapter)
            .join(Act, Act.id == Chapter.act_id)
            .join(Outline, Outline.id == Act.outline_id)
            .filter(Chapter.id == object_id, Outline.project_id == project_id)
            .first()
        )
    if object_type == "manuscript":
        return (
            db.query(Manuscript)
            .join(Chapter, Chapter.id == Manuscript.chapter_id)
            .join(Act, Act.id == Chapter.act_id)
            .join(Outline, Outline.id == Act.outline_id)
            .filter(Manuscript.id == object_id, Outline.project_id == project_id)
            .first()
        )
    return None


def _latest_version_data(db: Session, *, object_type: str, object_id: UUID, language: str) -> dict[str, Any]:
    latest = (
        db.query(ObjectVersion)
        .filter(ObjectVersion.object_type == object_type, ObjectVersion.object_id == object_id)
        .order_by(ObjectVersion.version_number.desc())
        .first()
    )
    if latest is None or not isinstance(latest.data, dict):
        return {}
    if isinstance(latest.data.get(language), dict):
        return latest.data[language]
    for v in latest.data.values():
        if isinstance(v, dict):
            return v
    return {}


def validate_tool_schema(args: dict[str, Any], tool_name: str, ctx: ValidationContext) -> ValidationResult:
    schema = _schema_for(tool_name, ctx)
    if schema is None:
        if tool_name.startswith("call_"):
            return valid_result()
        return invalid_result("validate_tool_schema", f"Unknown tool: {tool_name}")

    required = schema.parameters.get("required")
    properties = schema.parameters.get("properties")

    req_list = required if isinstance(required, list) else []
    prop_map = properties if isinstance(properties, dict) else {}

    for param in req_list:
        value = args.get(param)
        if value is None:
            return invalid_result("validate_tool_schema", f"Missing required parameter: {param}")
        if isinstance(value, str) and not value.strip():
            return invalid_result("validate_tool_schema", f"Required parameter cannot be empty: {param}")

    for key, value in args.items():
        schema_prop = prop_map.get(key)
        if not isinstance(schema_prop, dict):
            continue
        enum_vals = schema_prop.get("enum")
        if isinstance(enum_vals, list) and value not in enum_vals:
            return invalid_result(
                "validate_tool_schema",
                f"Invalid value for {key}: {value}. Expected one of: {', '.join(str(v) for v in enum_vals)}",
            )

    return valid_result()


def validate_tool_allowed_in_session(args: dict[str, Any], tool_name: str, ctx: ValidationContext) -> ValidationResult:
    _ = args
    if not ctx.allowed_tool_names:
        return valid_result()
    if tool_name in ctx.allowed_tool_names:
        return valid_result()
    return invalid_result("validate_tool_allowed_in_session", f"Tool not available in this session: {tool_name}")


def validate_sub_agent_call_input(args: dict[str, Any], tool_name: str, ctx: ValidationContext) -> ValidationResult:
    _ = ctx
    if not tool_name.startswith("call_"):
        return valid_result()
    value = args.get("input")
    if not isinstance(value, str) or not value.strip():
        return invalid_result("validate_sub_agent_call_input", 'call_* tools require a non-empty string field "input".')
    return valid_result()


def validate_type_mapping(args: dict[str, Any], tool_name: str, ctx: ValidationContext) -> ValidationResult:
    _ = ctx
    raw_type = args.get("type")
    if not isinstance(raw_type, str):
        return valid_result()

    if "story_object" in tool_name and raw_type not in STORY_TYPE_MAP:
        return invalid_result("validate_type_mapping", f"Unknown story object type: {raw_type}")

    if "_translation" in tool_name and raw_type not in ALL_OBJECT_TYPE_MAP:
        return invalid_result("validate_type_mapping", f"Unknown object type for translation: {raw_type}")

    return valid_result()


def validate_required_id(args: dict[str, Any], tool_name: str, ctx: ValidationContext) -> ValidationResult:
    _ = ctx
    if tool_name.startswith("create_"):
        return valid_result()
    value = _parse_required_field(args, "id")
    if value is None:
        return invalid_result("validate_required_id", f"Missing required id for {tool_name}")
    return valid_result()


def validate_patch_required_fields(args: dict[str, Any], tool_name: str, ctx: ValidationContext) -> ValidationResult:
    _ = ctx
    if not tool_name.startswith("patch_"):
        return valid_result()

    old_text = args.get("old")
    new_text = args.get("new")
    if not isinstance(old_text, str) or new_text is None:
        return invalid_result("validate_patch_required_fields", f"Missing required fields (old, new) for {tool_name}")

    if "manuscript" not in tool_name:
        field = args.get("field")
        if not isinstance(field, str) or not field.strip():
            return invalid_result("validate_patch_required_fields", f"Missing required field 'field' for {tool_name}")

    return valid_result()


def validate_rag_search_enabled(args: dict[str, Any], tool_name: str, ctx: ValidationContext) -> ValidationResult:
    _ = args
    if tool_name != "rag_search":
        return valid_result()
    if not ctx.rag_search_enabled:
        return invalid_result("validate_rag_search_enabled", "Embeddings are disabled (Settings > Search & Memory)")
    return valid_result()


async def validate_object_exists(args: dict[str, Any], tool_name: str, ctx: ValidationContext) -> ValidationResult:
    if tool_name.startswith("create_"):
        return valid_result()

    obj_id = _parse_required_field(args, "id")
    if obj_id is None:
        if tool_name == "patch_basic_info":
            # patch_basic_info has no id and is validated by project-level basic_info presence.
            basic = ctx.db.query(BasicInfo).filter(BasicInfo.project_id == ctx.run.project_id).first()
            if basic is None:
                return invalid_result("validate_object_exists", "No basic info found")
            return valid_result()
        return valid_result()

    object_type = _resolve_object_type(args, tool_name)
    if object_type is None:
        return invalid_result("validate_object_exists", f"Cannot determine object type for {tool_name}")

    try:
        object_uuid = UUID(obj_id)
    except (TypeError, ValueError):
        return invalid_result("validate_object_exists", f"Invalid object id format: {obj_id}")

    obj = _load_object_for_project(
        ctx.db,
        object_type=object_type,
        object_id=object_uuid,
        project_id=ctx.run.project_id,
    )
    if obj is None:
        return invalid_result("validate_object_exists", f"Object with id {obj_id} not found")
    return valid_result()


async def validate_patch_applicable(args: dict[str, Any], tool_name: str, ctx: ValidationContext) -> ValidationResult:
    if not tool_name.startswith("patch_"):
        return valid_result()

    old_text = args.get("old")
    new_text = args.get("new")
    if not isinstance(old_text, str) or new_text is None:
        return valid_result()

    if tool_name == "patch_basic_info":
        field = args.get("field")
        if not isinstance(field, str):
            return valid_result()
        basic = ctx.db.query(BasicInfo).filter(BasicInfo.project_id == ctx.run.project_id).first()
        if basic is None:
            return invalid_result("validate_patch_applicable", "No basic info found to patch")
        data = _latest_version_data(ctx.db, object_type="basic_info", object_id=basic.id, language=ctx.run.language)
        current = data.get(field)
        current_value = current if isinstance(current, str) else ""
        rr = apply_single_replacement(current_value, old_text, str(new_text))
        if not rr.success:
            return invalid_result(
                "validate_patch_applicable",
                f"Patch validation failed: [{field}] {rr.reason or rr.code or 'unknown'}",
            )
        return valid_result()

    object_type = _resolve_object_type(args, tool_name)
    obj_id = _parse_required_field(args, "id")
    if object_type is None or obj_id is None:
        return valid_result()

    try:
        object_uuid = UUID(obj_id)
    except (TypeError, ValueError):
        return valid_result()

    if object_type == "manuscript":
        if ctx.sidecar is None:
            return invalid_result("validate_patch_applicable", "SIDECAR_ERROR: sidecar client unavailable")

        data = _latest_version_data(ctx.db, object_type="manuscript", object_id=object_uuid, language=ctx.run.language)
        doc = data.get("doc")
        if not isinstance(doc, dict):
            return invalid_result("validate_patch_applicable", "MANUSCRIPT_EMPTY")

        try:
            markdown = await ctx.sidecar.doc_to_markdown(doc)
        except SidecarUnavailableError:
            return invalid_result("validate_patch_applicable", "SIDECAR_ERROR: unavailable")
        except SidecarConversionError as exc:
            return invalid_result("validate_patch_applicable", f"SIDECAR_ERROR: {exc}")

        rr = apply_single_replacement(markdown, old_text, str(new_text))
        if not rr.success:
            return invalid_result(
                "validate_patch_applicable",
                rr.code or rr.reason or "patch failed",
            )
        return valid_result()

    data = _latest_version_data(ctx.db, object_type=object_type, object_id=object_uuid, language=ctx.run.language)
    field = args.get("field")
    if not isinstance(field, str):
        return invalid_result("validate_patch_applicable", "Missing field for patch validation")

    current = data.get(field)
    current_value = current if isinstance(current, str) else ""
    rr = apply_single_replacement(current_value, old_text, str(new_text))
    if not rr.success:
        return invalid_result(
            "validate_patch_applicable",
            f"Patch validation failed: [{field}] {rr.reason or rr.code or 'unknown'}",
        )
    return valid_result()


async def validate_outline_parent_exists(args: dict[str, Any], tool_name: str, ctx: ValidationContext) -> ValidationResult:
    if tool_name != "create_outline_act":
        return valid_result()

    outline_id = _parse_required_field(args, "outlineId")
    if outline_id is None:
        return invalid_result("validate_outline_parent_exists", "Missing outlineId for create_outline_act")

    try:
        outline_uuid = UUID(outline_id)
    except (TypeError, ValueError):
        return invalid_result("validate_outline_parent_exists", "Invalid outlineId format")

    outline = ctx.db.query(Outline).filter(Outline.id == outline_uuid, Outline.project_id == ctx.run.project_id).first()
    if outline is None:
        return invalid_result("validate_outline_parent_exists", f"Parent outline not found: {outline_id}")

    return valid_result()


async def validate_act_parent_exists(args: dict[str, Any], tool_name: str, ctx: ValidationContext) -> ValidationResult:
    if tool_name != "create_outline_chapter":
        return valid_result()

    act_id = _parse_required_field(args, "actId")
    if act_id is None:
        return invalid_result("validate_act_parent_exists", "Missing actId for create_outline_chapter")

    try:
        act_uuid = UUID(act_id)
    except (TypeError, ValueError):
        return invalid_result("validate_act_parent_exists", "Invalid actId format")

    act = (
        ctx.db.query(Act)
        .join(Outline, Outline.id == Act.outline_id)
        .filter(Act.id == act_uuid, Outline.project_id == ctx.run.project_id)
        .first()
    )
    if act is None:
        return invalid_result("validate_act_parent_exists", f"Parent act not found: {act_id}")

    return valid_result()


GLOBAL_VALIDATORS: list[Validator] = [
    validate_tool_schema,
    validate_tool_allowed_in_session,
    validate_sub_agent_call_input,
]


TOOL_VALIDATORS: dict[str, list[Validator]] = {
    "create_story_object": [validate_type_mapping],
    "delete_story_object": [validate_type_mapping, validate_required_id, validate_object_exists],
    "create_outline": [],
    "delete_outline": [validate_required_id, validate_object_exists],
    "create_outline_act": [validate_outline_parent_exists],
    "delete_outline_act": [validate_required_id, validate_object_exists],
    "create_outline_chapter": [validate_act_parent_exists],
    "delete_outline_chapter": [validate_required_id, validate_object_exists],
    "replace_basic_info": [],
    "replace_guidelines": [validate_required_id, validate_object_exists],
    "replace_story_object": [validate_type_mapping, validate_required_id, validate_object_exists],
    "replace_manuscript": [validate_required_id, validate_object_exists],
    "replace_outline": [validate_required_id, validate_object_exists],
    "replace_outline_act": [validate_required_id, validate_object_exists],
    "replace_outline_chapter": [validate_required_id, validate_object_exists],
    "patch_basic_info": [validate_patch_required_fields, validate_object_exists, validate_patch_applicable],
    "patch_guidelines": [validate_required_id, validate_patch_required_fields, validate_object_exists, validate_patch_applicable],
    "patch_story_object": [validate_type_mapping, validate_required_id, validate_patch_required_fields, validate_object_exists, validate_patch_applicable],
    "patch_manuscript": [validate_required_id, validate_patch_required_fields, validate_object_exists, validate_patch_applicable],
    "patch_outline": [validate_required_id, validate_patch_required_fields, validate_object_exists, validate_patch_applicable],
    "patch_outline_act": [validate_required_id, validate_patch_required_fields, validate_object_exists, validate_patch_applicable],
    "patch_outline_chapter": [validate_required_id, validate_patch_required_fields, validate_object_exists, validate_patch_applicable],
    "read_story_object": [validate_type_mapping, validate_required_id, validate_object_exists],
    "read_outline": [validate_required_id, validate_object_exists],
    "read_manuscript": [validate_required_id, validate_object_exists],
    "rag_search": [validate_rag_search_enabled],
    "keyword_search": [],
}


async def run_validator_chain(validators: list[Validator], args: dict[str, Any], tool_name: str, ctx: ValidationContext) -> ValidationResult:
    for validator in validators:
        maybe = validator(args, tool_name, ctx)
        result = await maybe if inspect.isawaitable(maybe) else maybe
        if not result.valid:
            return result
    return valid_result()
